src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -1002,7 +1002,98 @@ class segment_meta_cstore::impl
         _write_buffer.erase(it);
     }
 
-    void insert(const segment_meta& m) {
+    bool can_be_inserted(const segment_meta& m) const {
+        // segment insertion is order dependant, so not all the sequence of
+        // insertions of the same set of segment_meta will result in the same
+        // end result. for example. given this manifest [0, 10][11, 20],
+        // inserting the segment [9, 12] is symptomatic of an underlying issue
+        // (no alignment whatsoever), but can't be rejected immediately. other
+        // edge cases can be rejected in the interest of keeping the manifest in
+        // a reasonable state: if the new segment is entirely inside an old
+        // segment it can be rejected. in this case [12, 20], [18,20] and [14,
+        // 17] should be rejected.
+        // in case of perfect overlap, the insertion can be done since it does
+        // not break the invariants, but the user of this class
+        // (partition_manifest) will apply further logic to restrict this case
+
+        // implementation, takes as parameters adapters to be able to operate on
+        // both _write_buffer and counter_col_t search m.base_offset. either
+        // it's an exact match, so the result can be inspected, or it's a
+        // segment starting after m.base_offset. if it's begin(), then m can be
+        // inserted (the container starts strictly after m) otherwise check the
+        // prev segment, as it might cover m.
+        constexpr static auto impl_can_be_inserted =
```

Comment:
should this be a separate method? then you can test it separately.

Reply:
done